### DISCUSSION TITLE
isDirty is not reset

### DISCUSSION THREAD

**Discussion by kaisiant** [react-hook-form/react-hook-form#4275] [Q&A]
1. create a new todo
2. edit the todo and save
Expect: save button disappear (isDirty = false)
Actual: save button still showing (isDirty = true)

My CSB:
https://codesandbox.io/s/frosty-pare-v9pc7?file=/src/EditTodos.tsx

**Comment by bluebill1049** [comment]
For field array, i think you should use this hook: https://react-hook-form.com/api/#useFieldArray

**Reply by kaisiant** [reply]
Me already using it

**Comment by bluebill1049** [comment]
<img width="1246" alt="Screen Shot 2021-02-25 at 12 11 40 pm" src="https://user-images.githubusercontent.com/10513364/109087975-a0034000-7762-11eb-949e-bcff8fdf80a6.png">

**Reply by kaisiant** [reply]
I guess you are not clear about the flow.
Type value into the input and press enter then you will see the array fields.
Take a look at the EditTodos.tsx

**Comment by leapful** [comment]
@bluebill1049 

You need to create a Todo first by press Enter and edit the added Todo.

<img width="549" alt="Screen Shot 2021-02-25 at 20 35 40" src="https://user-images.githubusercontent.com/18494222/109160984-1aef4980-77a9-11eb-9c9e-aee2a0c1e64e.png">

@bluebill1049 @kaisian97 

This issue is related to method "reset" in useForm() which forces reset form state. 

The CSB sample is currently using Controller therefore the force resetting will trigger clean up method "unregister" of useEffect inside useController hook.

https://github.com/react-hook-form/react-hook-form/blob/253f57d9d41952df31c2a5e02bbd65d16cbce517/src/useController.ts#L155

When performing "unregister", field ref of each input (contains its value) will be destroyed and form state "isDirty" is re-computed. 

https://github.com/react-hook-form/react-hook-form/blob/253f57d9d41952df31c2a5e02bbd65d16cbce517/src/useForm.ts#L764-L780

The method "isFormDirty" is currently comparing "fieldsRef" with "defaultValuesRef". At that time, default values will be the new changed input values  but fieldsRef has been destroyed in previous step "unregister", therefore both 2 values are never equal and form state "isDirty" is always marked as true.

@bluebill1049

Can we use "fieldArrayValuesRef" to compare with new "defaultValuesRef" for case re-compute form state "isDirty" when performing "unregistering" ?

**Reply by bluebill1049** [reply]
Thanks @leapful 

> Can we use "fieldArrayValuesRef" to compare with new "defaultValuesRef" for case re-compute form state "isDirty" when performing "unregistering" ?

Correct me if I am wrong, it may not be safe to compare just field array values with defaultValues.

**Reply by leapful** [reply]
@bluebill1049 Yes, just using field array values is not safe. Do we have any flag or data that is specified only for case "unregister" in order to only perform checking isDirty with field array values at that clean up time ?

**Reply by bluebill1049** [reply]
we don't have any flag, I think it's probably safer to just do a global dirty check against the `defaultValues`.
